src/modules/paymentProcessor/PP_Queue_v1.sol

```diff
@@ -0,0 +1,941 @@
+
+        // Check if order has started.
+        if (startTime_ > 0 && currentTime < startTime_) {
+            emit PaymentOrderTimingSkip(
+                orderId_,
+                _orders[orderId_].client,
+                SKIP_NOT_STARTED,
+                currentTime,
+                startTime_,
+                cliffEnd,
+                endTime_
+            );
+            return false;
+        }
+
+        // Check if cliff period has passed.
+        if (cliffPeriod_ > 0 && currentTime < cliffEnd) {
+            emit PaymentOrderTimingSkip(
+                orderId_,
+                _orders[orderId_].client,
+                SKIP_IN_CLIFF,
+                currentTime,
+                startTime_,
+                cliffEnd,
+                endTime_
+            );
+            return false;
+        }
+
+        // Check if order hasn't expired.
+        if (endTime_ > 0 && currentTime > endTime_) {
+            emit PaymentOrderTimingSkip(
+                orderId_,
+                _orders[orderId_].client,
+                SKIP_EXPIRED,
+                currentTime,
+                startTime_,
+                cliffEnd,
+                endTime_
+            );
+            return false;
+        }
+
+        return true;
+    }
+
+    /// @dev    Validate uint total amount input.
+    /// @param  amount_ Amount to validate.
+    /// @return valid_ True if uint is valid.
+    function _validTotal(uint amount_) internal pure returns (bool valid_) {
+        return amount_ != 0;
+    }
+
+    /// @dev    Validate whether the address is a valid payment receiver.
+    /// @param  receiver_ Address to validate.
+    /// @return validPaymentReceiver_ True if address is valid.
+    function _validPaymentReceiver(address receiver_)
+        internal
+        view
+        returns (bool validPaymentReceiver_)
+    {
+        return !(
+            receiver_ == address(0) || receiver_ == _msgSender()
+                || receiver_ == address(this)
+                || receiver_ == address(orchestrator())
+                || receiver_ == address(orchestrator().fundingManager().token())
+        );
+    }
+
+    /// @dev    Validates a payment token.
+    /// @param  token_ Token address to validate.
+    /// @return valid_ True if token is valid.
+    function _validPaymentToken(address token_)
+        internal
+        view
+        returns (bool valid_)
+    {
+        if (token_ == address(0)) {
+            revert Module__PP_Queue_InvalidToken(token_);
+        }
+
+        // Try to call balanceOf to verify it's an ERC20
+        try IERC20(token_).balanceOf(address(this)) returns (uint) {
+            return true;
+        } catch {
+            revert Module__PP_Queue_InvalidTokenImplementation(token_);
+        }
+    }
+
+    /// @dev    Validates a payment order.
+    /// @param  order_ The order to validate.
+    /// @return valid_ True if the order is valid.
+    function _validPaymentOrder(QueuedOrder memory order_)
```

Comment:
Where is this function called?

Reply:
Good eye. The `_validPaymentOrder` function should be called in `_addPaymentOrderToQueue` before creating the order, as this is where we receive new payment orders and should validate them before adding them to the queue.

Reply:
added it there now, btw

Reply:
Should we maybe call this function within the public PP function `validPaymentOrder()` as well? Instead of the check in the return value, as this function basically does the same if I'm not mistaken

Reply:
makes sense, done